plenum/server/primary_selector.py

```diff
@@ -54,14 +54,56 @@ def next_primary_node_name(self, instance_id, nodeReg=None):
 
         return name
 
-    def next_primary_replica_name(self, instance_id, nodeReg=None):
+    def next_primary_node_name_for_master(self, nodeReg=None):
+        if nodeReg is None:
+            nodeReg = self.node.nodeReg
+        rank = self._get_primary_id(self.viewNo, 0, len(nodeReg))
+        name = self.node.get_name_by_rank(rank, nodeReg=nodeReg)
+
+        logger.trace("{} selected {} as next primary node for master instance, "
+                     "viewNo {} with rank {}, nodeReg {}".format(
+                         self, name, self.viewNo, rank, nodeReg))
+        assert name, "{} failed to get next primary node name for master instance".format(self)
+
+        return name
+
+    def next_primary_node_name_for_backup(self, instance_id, nodeReg=None):
+        if nodeReg is None:
+            nodeReg = self.node.nodeReg
+        rank = self._get_primary_id(self.viewNo, instance_id, len(nodeReg))
+        name = self.node.get_name_by_rank(rank, nodeReg=nodeReg)
+
+        logger.trace("{} selected {} as next primary node for instId {}, "
+                     "viewNo {} with rank {}, nodeReg {}".format(
+                         self, name, instance_id, self.viewNo, rank, nodeReg))
+        assert name, "{} failed to get next primary node name".format(self)
+
+        return name
+
+    def next_primary_replica_name_for_master(self, nodeReg=None):
         """
         Returns name of the next node which is supposed to be a new Primary
         in round-robin fashion
         """
-        return Replica.generateName(
-            nodeName=self.next_primary_node_name(instance_id, nodeReg=nodeReg),
-            instId=instance_id)
+        name = self.next_primary_node_name_for_master(nodeReg)
+        return name, Replica.generateName(nodeName=name, instId=0)
+
+    def next_primary_replica_name_for_backup(self, instance_id, master_primary_rank,
+                                             primaries, nodeReg=None):
+        """
+        Returns name of the next node which is supposed to be a new Primary
```

Comment:
Please correct the description since it's not a real round robin, but it takes into account the other primaries.
Please mention that we return two values, not just name.

Reply:
Hmm... On my opinion it is pure round robin, the only thing is that we may skip some nodes, but it is round robin anyway.
As for primary for master replica - this is not round robin, I think. There is a formula that is not changed in scope of these changes:
(view_no + instance_id) % total_nodes
I really don't see round robin here.
As for returned two values - done.